File: _arch/algo/dynamic/robbery/tree/x.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */


 func rob(root *TreeNode) int {
    if root == nil{
        return 0
    }
    pair := robTree(root)
    return pair.max()
}


func robTree(node *TreeNode) Pair {
    if node == nil{
        return Pair{}
    }
    
    left := robTree(node.Left)
    right := robTree(node.Right)
    
    p := Pair{}
    p.now = node.Val + left.later + right.later
    p.later = left.max() + right.max()
    
    return p 
}

func max(a int, b int) int{
    if a > b {
        return a
    }
    return b 
}

type Pair struct {
    now int
    later int
}

func (p *Pair) max() int{
    return max(p.now, p.later)
}

File: _arch/algo/dynamic/robbery/tree/x.py
def rob(self, root: TreeNode) -> int:
    def robTree(node):
        """
        now: max money earned if input node is robbed
        later: max money earned if input node is not robbed
        """
        if not node:
            return 0, 0

        left, right = robTree(node.left), robTree(node.right)

        now = node.val + left[1] + right[1]

        later = max(left) + max(right)

        return now, later

    res = robTree(root)

    return max(res)
